test/Sentry.Tests/Internals/MemoryMonitorTests.cs

```diff
@@ -0,0 +1,220 @@
+    private HeapDumpTrigger AlwaysTrigger { get; } = (_, _) => true;
+
+    private class Fixture
+    {
+        public SentryOptions Options { get; set; } = new()
+        {
+            Dsn = ValidDsn,
+            Debug = true,
+            DiagnosticLogger = Substitute.For<IDiagnosticLogger>()
+        };
+
+        public Action<string> OnDumpCollected { get; set; } = _ => { };
+
+        public Action OnCaptureDump { get; set; } = null;
+
+        private const short ThresholdPercentage = 5;
+
+        public MemoryMonitor GetSut()
+        {
+            Options.DiagnosticLogger?.IsEnabled(Arg.Any<SentryLevel>()).Returns(true);
+            return new MemoryMonitor(Options, OnDumpCollected, OnCaptureDump);
+        }
+    }
+
+    private readonly Fixture _fixture = new();
+
+    [Fact]
+    public void Constructor_NoHeapdumpsConfigured_Throws()
+    {
+        // Arrange
+        _fixture.Options.HeapDumpOptions = null;
+
+        // Act
+        Assert.Throws<ArgumentException>(() => new MemoryMonitor(_fixture.Options, _fixture.OnDumpCollected));
+    }
+
+    [Fact]
+    public void CheckMemoryUsage_Debounced_DoesNotCapture()
+    {
+        // Arrange
+        _fixture.Options.EnableHeapDumps(
+            AlwaysTrigger,
+            Debouncer.PerApplicationLifetime(0) // always debounce
+        );
+        var dumpCaptured = false;
+        _fixture.OnCaptureDump = () => dumpCaptured = true;
+        using var sut = _fixture.GetSut();
+
+        // Act
+        sut.CheckMemoryUsage();
+
+        // Assert
+        dumpCaptured.Should().BeFalse();
+    }
+
+    [Fact]
+    public void CheckMemoryUsage_NotTriggered_DoesNotCapture()
+    {
+        // Arrange
+        _fixture.Options.EnableHeapDumps(
+            NeverTrigger,
+            Debouncer.PerApplicationLifetime(int.MaxValue) // never debounce
+        );
+        var dumpCaptured = false;
+        _fixture.OnCaptureDump = () => dumpCaptured = true;
+        using var sut = _fixture.GetSut();
+
+        // Act
+        sut.CheckMemoryUsage();
+
+        // Assert
+        dumpCaptured.Should().BeFalse();
+    }
+
+    [Fact]
+    public void CheckMemoryUsage_TriggeredNotDebounced_Captures()
+    {
+        // Arrange
+        _fixture.Options.EnableHeapDumps(
+            AlwaysTrigger,
+            Debouncer.PerApplicationLifetime(int.MaxValue) // never debounce
+        );
+        var dumpCaptured = false;
+        _fixture.OnCaptureDump = () => dumpCaptured = true;
+        using var sut = _fixture.GetSut();
+
+        // Act
+        sut.CheckMemoryUsage();
+
+        // Assert
+        dumpCaptured.Should().BeTrue();
+    }
+
+    [Fact]
+    public void CaptureMemoryDump_DisableFileWrite_DoesNotCapture()
+    {
+        // Arrange
+        _fixture.Options.EnableHeapDumps(AlwaysTrigger);
+        _fixture.Options.DisableFileWrite = true;
+        using var sut = _fixture.GetSut();
+
+        // Act
+        sut.CaptureMemoryDump();
```

Comment:
Any tests where `CaptureMemoryDump` throw? More concerned about the caller actually. Since we don't want to crash the app, do we handle every Task we create?

Reply:
The Task that calls that is ultimately this:
https://github.com/getsentry/sentry-dotnet/blob/58d9e556a6ab4bb8d760aab9dd5fed321e0d421a/src/Sentry/Internal/MemoryMonitor.cs#L40-L46

I played around with various things and eventually settled on a FireAndForget with a Continuation. The alternative of holding a reference to the task wasn't ideal because you only become aware of any errors when you await the task... which here would have to happen when disposing of the MemoryMonitor (application shutdown). I wanted to log any errors as soon as they occur however.

I've refactored the `GarbageCollectionMonitor` and added some `GarbageCollectionMonitorTests` just to make sure cancellation happens as we expect (without bubbling any exceptions up that might disrupt the calling code). 

I think ultimately what you're wanting to test is [that task continuation](https://github.com/getsentry/sentry-dotnet/blob/58d9e556a6ab4bb8d760aab9dd5fed321e0d421a/src/Sentry/Internal/MemoryMonitor.cs#L40-L46) but the fire and forget nature of that code makes it difficult to test. I did test it manually in a console app, when I was investigating how to detect any non-cancellation errors as soon as they happen.